### PR TITLE
Don't use the broken overflow builtins for pointers

### PR DESCRIPTION
**Pull request by danielsn**
### Description of changes: 

Currently, we emit `__builtin_overflow_p` operations for pointer arithmetic.  This is basically broken, since CBMC converts to integers and does integer arithmetic under the hood, which is not what we want. 

### Resolved issues:

Initial step towards addressing https://github.com/model-checking/kani/issues/786

### Call-outs:

1. This relies on `--pointer-overflow-check` being used on CBMC for soundness.  We currently do so, but it would be better to have a `__pointer_overflow` builtin we could use when we need this behaviour. https://github.com/diffblue/cbmc/issues/6627

2. Also fixed some confusing missing brackets in the typechecking code for binops.

### Testing: 

* How is this change tested? Existing regression tests. 

* Is this a refactor change? No

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: src/kani-compiler/cbmc/src/goto_program/expr.rs [hunk 1]
```diff
@@ -1205,7 +1203,18 @@ impl Expr {
     /// `ArithmeticOverflowResult r; >>>r.overflowed = builtin_add_overflow(self, e, &r.result)<<<`
     pub fn add_overflow(self, e: Expr) -> ArithmeticOverflowResult {
         let result = self.clone().plus(e.clone());
-        let overflowed = self.add_overflow_p(e);
+        // FIXME: https://github.com/model-checking/kani/issues/786
+        // Overflow checking on pointers is hard to do at the IREP level.
+        // Instead, we rely on `--pointer-overflow-check` in CBMC.
+        let overflowed = if self.typ.is_pointer() {
```
Comment (adpaco-aws) [inline src/kani-compiler/cbmc/src/goto_program/expr.rs:1209]: Don't we have to check if `e` is a pointer here?

File: src/kani-compiler/cbmc/src/goto_program/expr.rs [hunk 2]
```diff
@@ -851,13 +852,10 @@ impl Expr {
...
-                (lhs.typ == rhs.typ && (lhs.typ.is_pointer() || lhs.typ.is_numeric()))
-                    || (lhs.typ.is_pointer() && rhs.typ.is_integer())
-            }
-            OverflowMult | OverflowPlus => {
-                (lhs.typ == rhs.typ && lhs.typ.is_integer())
-                    || (lhs.typ.is_pointer() && rhs.typ.is_integer())
+            // While one can technically use these on pointers, the result is treated as an integer.
+            // This is almost never what you actually want.
+            OverflowMinus | OverflowMult | OverflowPlus => {
+                lhs.typ.is_integer() && rhs.typ.is_integer()
```
Comment (adpaco-aws) [inline src/kani-compiler/cbmc/src/goto_program/expr.rs:858]: Suggestion: Follow the same style as in other arms.
```suggestion
                (lhs.typ == rhs.typ && lhs.typ.is_integer())
```

File: src/kani-compiler/cbmc/src/goto_program/expr.rs [hunk 3]
```diff
@@ -851,13 +852,10 @@ impl Expr {
             // Floating Point Equalities
             IeeeFloatEqual | IeeeFloatNotequal => lhs.typ == rhs.typ && lhs.typ.is_floating_point(),
             // Overflow flags
-            OverflowMinus => {
-                (lhs.typ == rhs.typ && (lhs.typ.is_pointer() || lhs.typ.is_numeric()))
```
Comment (adpaco-aws) [inline src/kani-compiler/cbmc/src/goto_program/expr.rs:855]: Good to restrict this case to integers because I do not think we were handling floating point numbers anyway.

**Comment by celinval** [comment]
I don't have a lot of context here, but I thought that was the goal of these checks. They were complementary to CBMC pointer checks. These were just making sure the arithmetic operations would not overflow.

One could argue that they are redundant but I don't know if broken is how I would describe them.

**Comment by celinval** [comment]
Can you please add a test that would trigger the overflow to ensure the verification will still fail due to other checks?

File: src/kani-compiler/cbmc/src/goto_program/expr.rs (hunk 2 above)
Comment (danielsn) [reply src/kani-compiler/cbmc/src/goto_program/expr.rs:858]: done

File: src/kani-compiler/cbmc/src/goto_program/expr.rs (hunk 1 above)
Comment (danielsn) [reply src/kani-compiler/cbmc/src/goto_program/expr.rs:1209]: Technically, that we'd notice any cases cause type-checking would fail. But making the test is clearer.

**Comment by danielsn** [comment]
> Can you please add a test that would trigger the overflow to ensure the verification will still fail due to other checks?

Done

**Comment by danielsn** [comment]
> I don't have a lot of context here, but I thought that was the goal of these checks. They were complementary to CBMC pointer checks. These were just making sure the arithmetic operations would not overflow.
> 
> One could argue that they are redundant but I don't know if broken is how I would describe them.

The issue is that the semantics of the arithmetic operation being done, and the semantics of the check, don't line up.  The arithmetic operation is doing *pointer* arithmetic, while the overflow check is doing *integer* arithmetic. So its possible for the actual addition to overflow, but the overflow check to report no overflow.

**Comment by celinval** [comment]
> > I don't have a lot of context here, but I thought that was the goal of these checks. They were complementary to CBMC pointer checks. These were just making sure the arithmetic operations would not overflow.
> > One could argue that they are redundant but I don't know if broken is how I would describe them.
> 
> The issue is that the semantics of the arithmetic operation being done, and the semantics of the check, don't line up. The arithmetic operation is doing _pointer_ arithmetic, while the overflow check is doing _integer_ arithmetic. So its possible for the actual addition to overflow, but the overflow check to report no overflow.

Can we account for the type size?

**Comment by celinval** [comment]
I believe this change is currently blocked waiting for a CBMC fix to be released. @danielsn Please let us know once this is ready for review again. Cheers!
